File: src/main/java/ash/java/graphql/fields/MovieSearchSchema.java
package ash.java.graphql.fields;

import ash.java.graphql.data.SearchDao;
import ash.java.graphql.types.movie.MovieType;
import graphql.schema.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import static graphql.Scalars.*;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;

@Service
public class MovieSearchSchema implements FieldProducer {

    private SearchDao searchDao;

    @Autowired
    public MovieSearchSchema(SearchDao searchDao) {
        this.searchDao = searchDao;
    }

    private GraphQLFieldDefinition movieSearchFieldDefinition = newFieldDefinition()
            .type(new GraphQLList(new MovieType().getGraphQlType()))
            .name("movieSearch")
            .argument(arg -> arg.name("query").type(new GraphQLNonNull(GraphQLString)))
            .argument(arg -> arg.name("language").type(GraphQLString))
            .argument(arg -> arg.name("page").type(GraphQLInt))
            .argument(arg -> arg.name("includeAdult").type(GraphQLBoolean))
            .argument(arg -> arg.name("region").type(GraphQLString))
            .argument(arg -> arg.name("year").type(GraphQLInt))
            .argument(arg -> arg.name("primaryReleaseYear").type(GraphQLInt))
            .dataFetcher(env -> searchDao.searchMoviesWithMultipleParameters(env.getArguments()))
            .build();

    public GraphQLFieldDefinition getFieldDefinition() {
        return movieSearchFieldDefinition;
    }
}

File: src/main/kotlin/ash/kotlin/graphql/fields/MovieSearchSchema.kt
package ash.kotlin.graphql.fields

import ash.kotlin.graphql.data.SearchDao
import ash.kotlin.graphql.types.movie.MovieType
import graphql.Scalars.*
import graphql.schema.*

class MovieSearchSchema constructor(private val dao: SearchDao) : FieldProducer
{
    override fun getFieldDefinition(): GraphQLFieldDefinition
    {
        return GraphQLFieldDefinition.newFieldDefinition()
                .type(GraphQLList(MovieType().getGraphQlType()))
                .name("movieSearch")
                .argument { arg -> arg.name("query").type(GraphQLNonNull(GraphQLString)) }
                .argument { arg -> arg.name("language").type(GraphQLString) }
                .argument { arg -> arg.name("page").type(GraphQLInt) }
                .argument { arg -> arg.name("includeAdult").type(GraphQLBoolean) }
                .argument { arg -> arg.name("region").type(GraphQLString) }
                .argument { arg -> arg.name("year").type(GraphQLInt) }
                .argument { arg -> arg.name("primaryReleaseYear").type(GraphQLInt) }
                .dataFetcher { env -> dao.searchMoviesWithMultipleParameters(env.arguments) }
                .build()
    }
}